awx/main/models/unified_jobs.py

```diff
@@ -1512,3 +1514,52 @@ def log_lifecycle(self, state, blocked_by=None):
         else:
             msg = f"{self._meta.model_name}-{self.id} {state.replace('_', ' ')}"
         logger_job_lifecycle.debug(msg, extra=extra)
+
+#not doing the recursive logic right now , but just trying to make sure it returns something , the logic is basically what is used in the UI.
+    @property
+    def launched_by(self):
+        
+        # if the job is some maintainance task
+        #sdb.set_trace()
+        if hasattr(self,'job_type'):
+            if self.job_type=="cleanup_activity_stream":
+                return None
+        else:
+            return None
+            
+        
+        #sdb.set_trace()
+        #checking if the current job was started by a workflow job.
+       
+        ancestor_job=self.ancestor_job
+
+        if hasattr(ancestor_job,'launch_type'):
+            if ancestor_job.launch_type=="manual":
+                return ancestor_job.created_by
+            elif ancestor_job.launch_type=="scheduled":
+                return ancestor_job.schedule
+            else:
+                return None
+        else:
+            return None
+
+    @property
+    def ancestor_job(self):
+        
+
+        if hasattr(self,'get_workflow_job()'):
```

Comment:
Two things,
- `hasattr` is not going to work properly with the parentheses.  `hasattr(self, 'get_workflow_job')`.  Similarly on line 1557.
- guarding against the method not being there isn't necessary.  `get_workflow_job` is defined on `UnifiedJob`, so will be available for all job types.